### ISSUE TITLE
Offset size too big

### ISSUE THREAD

**Issue by CMajeri** [ton-blockchain/ton#77]
Hello, 
I've spotted inconsistencies with regards to the offset size when serializing bags of cells.

I'm playing around with the fift interpreter to generate a cell with some smart contract code (the new-wallet one, incidentally).
The cell is 83 bytes long (including the header)
I'm attempting to serialize it into a bag, since it is the only cell in the bag, the "largest possible offset" is simply this length. to represent 83, we need ~7bits, so just one byte.
However, when I serialize it (with fift) and look through it manually, the `off_bytes` (which I assume is "the number of bytes necessary to store the max offset length") is set to 4.
Looking at the `tot_cell_size`, I see what I expected : `00000053`. 
So 3 null bytes, and the length of my only cell.

Looking through the code, I noticed that this value is computed here:
https://github.com/ton-blockchain/ton/blob/master/crypto/vm/boc.cpp#L387

This line seems bugged to me, and should be 
```
  while (max_offset >= (1ULL << (os << 8))) {
    os++;
  }
```

I feel similarly about the computing of rs, where a `cell_count` of 9 would result in a 2 bytes "size".

This *shouldn't* break anything in either case, I'm just wondering why waste space during the serialization of boc's.

Below is a small snippet of what I serialized (with the fift interpreter)
```
CS{Cell{00a2ff0020dd2082014c97ba9730ed44d0d70b1fe0a4f260810200d71820d70b1fed44d0d31fd3ffd15112baf2a122f901541044f910f2a2f80001d31f3120d74a96d307d402fb00ded1a4c8cb1fcbffc9ed54} bits: 0..648; refs: 0..0}

B5EE9C72
  01 = b00000001
    has_idx = b0
    has_cache_bits = b0
    flags = b00
    size = b001
  04 =
    off_bytes = 04
  01 =
    cells = 01
  01 =
    roots = 01
  00 = 
    absent = 00
  00000053 = (off_bytes * 8)
    tot_cells_size = 00000053
  00 = (roots * 8)
    root_list = [00]
  00A2FF0020DD2082014C97BA9730ED44D0D70B1FE0A4F260810200D71820D70B1FED44D0D31FD3FFD15112BAF2A122F901541044F910F2A2F80001D31F3120D74A96D307D402FB00DED1A4C8CB1FCBFFC9ED54 =
    cell_0

B5EE9C720104010100000000530000A2FF0020DD2082014C97BA9730ED44D0D70B1FE0A4F260810200D71820D70B1FED44D0D31FD3FFD15112BAF2A122F901541044F910F2A2F80001D31F3120D74A96D307D402FB00DED1A4C8CB1FCBFFC9ED54
```
and how I serialized it:
```
 // Create new simple wallet
<{ SETCP0 DUP IFNOTRET // return if recv_internal
   DUP 85143 INT EQUAL IFJMP:<{ // "seqno" get-method
     DROP c4 PUSHCTR CTOS 32 PLDU  // cnt
   }>
   INC 32 THROWIF  // fail unless recv_external
   512 INT LDSLICEX DUP 32 PLDU   // sign cs cnt
   c4 PUSHCTR CTOS 32 LDU 256 LDU ENDS  // sign cs cnt cnt' pubk
   s1 s2 XCPU            // sign cs cnt pubk cnt' cnt
   EQUAL 33 THROWIFNOT   // ( seqno mismatch? )
   s2 PUSH HASHSU        // sign cs cnt pubk hash
   s0 s4 s4 XC2PU        // pubk cs cnt hash sign pubk
   CHKSIGNU              // pubk cs cnt ?
   34 THROWIFNOT         // signature mismatch
   ACCEPT
   SWAP 32 LDU NIP 
   DUP SREFS IF:<{
     // 3 INT 35 LSHIFT# 3 INT RAWRESERVE    // reserve all but 103 Grams from the balance
     8 LDU LDREF         // pubk cnt mode msg cs
     s0 s2 XCHG SENDRAWMSG  // pubk cnt cs ; ( message sent )
   }>
   ENDS
   INC NEWC 32 STU 256 STU ENDC c4 POPCTR
}>c // >libref
dup <s .s drop 
dup boc>B .s drop
0 halt
```

**Comment by CMajeri** [comment]
Nevermind, I'm completely wrong!
But can someone explain why we need this ?:
```
    if (rs > 2 || os > 4) {
      rs = std::max(3, rs);
      os = 8;
      data_bytes_adj = data_bytes + (unsigned long long)int_refs * rs + hashes;
    } else {
      os = 4;
    }
```

**Comment by ton-blockchain** [comment]
Yes, you probably wanted to say
`   while (max_offset >= (1ULL << (os * 8))) ...`
not
` while (max_offset >= (1ULL << (os << 8))) ...`

The first form is equivalent to `while (max_offset >= (1ULL << (os << 3))`, as in the sources.

**Comment by CMajeri** [comment]
Yeah I noticed that after, I was off on that, didn't want to edit away my mistake :)

What's up with the `if` at the end of the function though?

**Comment by ton-blockchain** [comment]
That `if` appears to be a rudiment of older serialization schemes, which can still be found in 5.3.9 of the [Blockchain documentation](https://test.ton.org/tblkch.pdf). The currently supported layout is in [boc.tlb](https://github.com/ton-blockchain/ton/blob/master/crypto/tl/boc.tlb); only the `serialized_boc#b5ee9c72` form is used for serialization now.

Thanks for bringing this issue to our attention. It is not a serious bug, but it still led to some unnecessary waste of network traffic.

We'll study this issue further and probably just remove this `if`, once we check that this does not break anything.

**Comment by CMajeri** [comment]
Yeah, that's what I thought too.
It seemed at odd with the boc.tlb file.

Thanks!

**Comment by ton-blockchain** [comment]
Fixed in commit dd745485e26cf581019c2b8c3013d4c14ec0473e
